contrib/epee/include/net/abstract_tcp_server2.inl

```diff
@@ -81,860 +81,1019 @@ namespace net_utils
+    if (state.protocol.wait_release)
+      return;
+    if (state.socket.connected) {
+      ec_t ec;
+      connection_basic::socket_.next_layer().shutdown(
+        socket_t::shutdown_both,
+        ec
+      );
+      connection_basic::socket_.next_layer().close(ec);
+      state.socket.connected = false;
+    }
+    state.status = status_t::WASTED;
   }
-  //---------------------------------------------------------------------------------
-  template<class t_protocol_handler>
-  bool connection<t_protocol_handler>::call_run_once_service_io()
+
+  template<typename T>
+  bool connection<T>::send(byte_slice_t message)
   {
-    TRY_ENTRY();
-    if(!m_is_multithreaded)
-    {
-      //single thread model, we can wait in blocked call
-      size_t cnt = GET_IO_SERVICE(socket()).run_one();
-      if(!cnt)//service is going to quit
-        return false;
-    }else
-    {
-      //multi thread model, we can't(!) wait in blocked call
-      //so we make non blocking call and releasing CPU by calling sleep(0); 
-      //if no handlers were called
-      //TODO: Maybe we need to have have critical section + event + callback to upper protocol to
-      //ask it inside(!) critical region if we still able to go in event wait...
-      size_t cnt = GET_IO_SERVICE(socket()).poll_one();
-      if(!cnt)
-        misc_utils::sleep_no_w(1);
+    lock_guard_t guard(state.lock);
+    if (state.status != status_t::RUNNING || state.socket.wait_handshake)
+      return false;
+    auto random_delay = []{
+      using engine = std::mt19937;
+      std::random_device dev;
+      std::seed_seq::result_type rand[
+        engine::state_size // Use complete bit space
+      ]{};
+      std::generate_n(rand, engine::state_size, std::ref(dev));
+      std::seed_seq seed(rand, rand + engine::state_size);
+      engine rng(seed);
+      return std::chrono::milliseconds(
+        std::uniform_int_distribution<>(5000, 6000)(rng)
+      );
+    };
+    if (state.data.write.wait_consume ||
+      state.data.write.queue.size() > ABSTRACT_SERVER_SEND_QUE_MAX_COUNT
+    )
+      state.condition.wait_for(state.lock, random_delay(),
+        [this]{
+          return (
+            state.status != status_t::RUNNING ||
+            (
+              !state.data.write.wait_consume  &&
+              state.data.write.queue.size() <=
+                ABSTRACT_SERVER_SEND_QUE_MAX_COUNT
+            )
+          );
+        }
+      );
+    if (state.status != status_t::RUNNING)
+      return false;
+    constexpr size_t CHUNK_SIZE = 32 * 1024;
+    if (connection_type == e_connection_type_RPC ||
+      message.size() <= 2 * CHUNK_SIZE
+    ) {
+      state.data.write.queue.emplace_front(std::move(message));
+      start_write();
+      return true;
     }
-    
+    state.data.write.wait_consume = true;
+    while (!message.empty()) {
+      if (state.data.write.queue.size() > ABSTRACT_SERVER_SEND_QUE_MAX_COUNT) {
+        state.condition.wait_for(state.lock, random_delay(),
```

Comment:
This is still a DoS case - the thread is still blocked.This should be changed to a drop (this is one of the reasons why I am pushing for a piecemeal approach ...).

Reply:
Both connection and sender should be asynchronous to remove this problem.
It's partially solved now, as much as possible without changing the rest of the code.

Reply:
This queue will be removed from connection later, once everything else is asynchronous.

Reply:
You can't remove a queue, or at least you shouldn't. Because `post`ing to a strand is still a queue.

AFAIK there are 3 choices: (1) close connection (drop data), (2) block a thread to wait, or (3) unbounded queue. I think this needs to transition to (1) if an entire re-write is being done.

Reply:
I can, will and it's needed. But it isn't easy.

Reply:
`post` and any other forward will be replaced with asynchronous generator

Reply:
Asynchronous generator? You mean asynchronously generate bytes piecemeal? That's just a more complicated outgoing buffer. This has to be a drop or block.

Reply:
What are the sources of data for outgoing queue ?
1. Synchronous forward of notifications with txs/blocks which can be made asynchronous with fetching from local blockchain storage.
2. Requests/responses required by explicit synchronization which can be generated asynchronously as soon as outgoing/incoming fixed size buffer is free.

In general, daemon sends and receives only verified data that is stored/loaded to/from blockchain db.
This is the only available unbounded buffer to store data.

Reply:
I already know exactly what you are proposing, and I am pointing out that some p2p messages require a response, and a queue of asynchronous events is still a queue. The commentary by the ZeroMQ developers on Pub/Sub seems appropriate here - if the read side is behind eventually you have to drop data and/or the connection.

All of these changes don't really change the behavior of the existing problem. Multiple threads can get stuck busy waiting for a queue to clear. The previous method had the threads waiting indefinitely at a lock, whereas here they wait indefinitely in conditional variables.

Reply:
> This is still a DoS case - the thread is still blocked.This should be changed to a drop (this is one of the reasons why I am pushing for a piecemeal approach ...).

I didn't understand that you were talking about any action after failed conditional sleep.
I've added missing `terminate(); return false`.

Reply:
> All of these changes don't really change the behavior of the existing problem. Multiple threads can get stuck busy waiting for a queue to clear. The previous method had the threads waiting indefinitely at a lock, whereas here they wait indefinitely in conditional variables.

AFAICT this approach is a marked improvement over the current code:

1. Reads/writes are no longer waiting on `handle_recv` or callbacks to complete (which execute in the newly introduced 2nd `strand_`), therefore they can process quicker and clear from the strand quicker.
2. It uses a conditional wait that can unblock as soon as the queue is less than the max.

Without a demonstrable DoS on this new code as is which is evidently harder to DoS, I think the code is merge-able as is, and worth continuing to improve, which the PR author intends to do building from the structure of this PR's code

Additionally there is a risk in dropping the connection here, which is that it can feasibly harm honest usage of the daemon. It's going to take more work to dig deeper to make this entire section air tight to hit the optimal sweet spot imo. For now, this code is an improvement moving in a better direction.